src/tests/aes.rs

```diff
@@ -746,6 +746,320 @@ fn test_aes_operations() {
         assert_eq!(value, data);
     }
 
+    {
+        /* GCM via AEAD MessageEncrypt/MessageDecrypt API */
+        let mut mechanism: CK_MECHANISM = CK_MECHANISM {
+            mechanism: CKM_AES_GCM,
+            pParameter: std::ptr::null_mut(),
+            ulParameterLen: 0,
+        };
+
+        let ret = fn_message_encrypt_init(session, &mut mechanism, handle);
+        assert_eq!(ret, CKR_OK);
+
+        /* IV needs to be of size 12 for the test to work in FIPS mode as well */
+        let iv = "BA0987654321";
+        let aad = "AUTH ME";
+        let mut tag = [0u8; 4];
+        let mut param = CK_GCM_MESSAGE_PARAMS {
+            pIv: iv.as_ptr() as *mut CK_BYTE,
+            ulIvLen: iv.len() as CK_ULONG,
+            ulIvFixedBits: 0,
+            ivGenerator: CKG_NO_GENERATE,
+            pTag: tag.as_mut_ptr(),
+            ulTagBits: (tag.len() * 8) as CK_ULONG,
+        };
+
+        let ret = fn_encrypt_message_begin(
+            session,
+            void_ptr!(&mut param),
+            sizeof!(CK_GCM_MESSAGE_PARAMS),
+            byte_ptr!(aad.as_ptr()),
+            aad.len() as CK_ULONG,
+        );
+        assert_eq!(ret, CKR_OK);
+
+        /* Stream mode, so arbitrary data size and matching output */
+        let data = "01234567";
+        let enc: [u8; 8] = [0; 8];
+        let mut enc_len = enc.len() as CK_ULONG;
+        let ret = fn_encrypt_message_next(
+            session,
+            void_ptr!(&mut param),
+            sizeof!(CK_GCM_MESSAGE_PARAMS),
+            data.as_ptr() as *mut CK_BYTE,
+            (data.len() - 1) as CK_ULONG,
+            enc.as_ptr() as *mut _,
+            &mut enc_len,
+            0,
+        );
+        assert_eq!(ret, CKR_OK);
+        assert_eq!(enc_len as usize, data.len() - 1);
+
+        enc_len = 1 as CK_ULONG;
+        let ret = fn_encrypt_message_next(
+            session,
+            void_ptr!(&mut param),
+            sizeof!(CK_GCM_MESSAGE_PARAMS),
+            unsafe { data.as_ptr().offset(7) } as *mut CK_BYTE,
+            1 as CK_ULONG,
+            unsafe { enc.as_ptr().offset(7) } as *mut _,
+            &mut enc_len,
+            CKF_END_OF_MESSAGE,
+        );
+        assert_eq!(ret, CKR_OK);
+        assert_eq!(enc_len, 1);
+
+        /* test that we can get correct indicators based on inputs */
+        assert_eq!(check_validation(session, 0), true);
```

Comment:
this should return valid fips check, I think.

Reply:
no because the IV is passed in.